Support anonymous tokens

jhannes opened the pull request:
This is a pull request for supporting anonymous tokens as per https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Documentation/issues/4. The change replaces the use of AuthorizationAttribute on DiagnosticsKeyControllerV3.UploadDiagnosticsKeys with a new attribute UploadKeysAuthorizationAttribute.

The new attribute is an extension of the previous one which also supports "Anonymous " values in the "Authorization"-header. If an Anonymous authorization is provided, the code verifies the token according to https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Documentation/issues/4, otherwise, the same code as before is used.

The existing AuthorizationAttribute is retained to minimize the change

jhannes committed:
Add UploadKeysAuthorizationAttribute which supports validating anonymous tokens

jhannes committed:
Easier local debugging

matishadow commented on DIGNDB.App.SmitteStop.API/Attributes/UploadKeysAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,82 @@
+{
+    public class UploadKeysAuthorizationAttribute : ActionFilterAttribute
+    {
+        private readonly AppSettingsConfig _appSettingsConfig;
```
`_appSettingsConfig` field is not used, please remove it if you don't need it.

jhannes replied:
I will it both here and from AuthorizationAttributes to keep the files easy to compare

matishadow commented on DIGNDB.App.SmitteStop.API/Attributes/UploadKeysAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,82 @@
+
+namespace DIGNDB.App.SmitteStop.API.Attributes
+{
+    public class UploadKeysAuthorizationAttribute : ActionFilterAttribute
```
Don't you think more generic name would be better?
Something not tightly coupled to uploading keys.
Maybe `AnonymousAuthorizationAttribute`? 

It's just a suggestion, something to reconsider.

matishadow commented on DIGNDB.App.SmitteStop.API/Attributes/UploadKeysAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,82 @@
+                    return;
+
+                }
+                else if (authHeader != null && authHeader.StartsWith("Anonymous ")) 
```
This `UploadKeysAuthorizationAttribute` is in my opinion pretty similar to `AuthorizationAttribute`.
I think you could avoid some code duplication by using inheritance or composition pattern. 

For example in this class you could only have this condition:
```
else if (authHeader != null && authHeader.StartsWith("Anonymous ")) 
```
and code inside it.

If this condition is not met then you should direct the execution to `AuthorizationAttribute` class.

jhannes replied:
It's a copy on purpose and many of the issues here are result of copying AuthorizationAttribute. By copying, I minimize the risk that the change has unintended consequences by reducing coupling. But when the change has been fully tested, this duplication should be removed.

matishadow replied:
Okay I agree, but when is it going to be fully tested?

jhannes replied:
When FHI decides to turn on anonymous tokens in production

matishadow replied:
I'll resolve this comment and create an issue to remove this duplication to not forget about it!

matishadow commented on DIGNDB.App.SmitteStop.API/ContainerRegistration.cs:
```diff
@@ -67,7 +66,7 @@ public static IServiceCollection AddAPIDependencies(this IServiceCollection serv
                 var xmlPath = Path.Combine(AppContext.BaseDirectory, xmlFile);
                 c.IncludeXmlComments(xmlPath);
             });
-            services.AddSingleton(new AuthOptions(environment.IsDevelopment()));
```
I don't think this should be removed.

Removing it is okay for Upload keys endpoint (provided you have the mocked jwk endpoint). 
But it's not for other endpoints.

By removing this you will make debugging Download keys harder (because it uses `MobileAuthorization` which also uses `AuthOptions`). 

I think the better solution would be to not use `AuthOptions` when Uploading keys.
![image](https://user-images.githubusercontent.com/9083085/102344279-f2bebf80-3f9b-11eb-8c6a-e283c7da32be.png)
(remove the selected line).

But restore this:
```
services.AddSingleton(new AuthOptions(environment.IsDevelopment()));
```
so other endpoints are not affected.

jhannes replied:
Sure. But I recommend reducing differences between development and production environments. I didn't examine the details, but what's needed to get MobileAuthorization working on local development?

matishadow replied:
I agree with reducing differences! 

`MobileAuthorization` needs this value to be set in `appsettings.json`:
![image](https://user-images.githubusercontent.com/9083085/102352012-9d88ab00-3fa7-11eb-96db-f44259e5aa41.png)

This value is a static secret shared between Mobile App and Backend.
Unfortunately it is encrypted using `Windows Data Protection API`, which makes it very hard to have 'a development value'.
It is hard because the encryption is being done using Windows Account credentials, so each user needs to have a different value there.

Because of this inconvenience we decided to disable the MobileAuthorization check on development environments.

jhannes replied:
I think more research on the part of the dev team would be prudent here, but for now, I'll revert this change.

matishadow commented on DIGNDB.App.SmitteStop.API/appsettings.Development.json:
```diff
@@ -4,9 +4,22 @@
+      "SupportedAlgorithm": "RS256",
+      "Issuer": "http://localhost:5001"
+    },
+    "JwkUrl": "http://localhost:5001/.well-known/openid-configuration/jwks"
```
Should this endpoint `http://localhost:5001/.well-known/openid-configuration/jwks` be running automatically with the use of the new package `AnonymousTokens.Server`?
If not then it'd be nice to mention how to set it up. 
Maybe we should consider adding another simple project which would start up this mock?

jhannes replied:
This is started by https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Verification/

matishadow replied:
Great!
Could you please add a comment in the development json file with this link so people will know where to look?

jhannes replied:
I don't think anything is needed to do here. The README.md is currently empty, but as we finalize the questions on Slack, I intent to write what's needed for a full local development environment,

matishadow replied:
I'll resolve this and create and issue to not forget about it.

matishadow commented on DIGNDB.App.SmitteStop.API/Attributes/UploadKeysAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,82 @@
+                }
+                else if (authHeader != null && authHeader.StartsWith("Anonymous ")) 
+                {
+                    var _ecParameters = CustomNamedCurves.GetByOid(X9ObjectIdentifiers.Prime256v1);
```
Local variables i.e. `_ecParameters` should not be prefixed with underscore `_`.

jhannes replied:
Whoops. Copy paste error. Will fix.

matishadow commented on DIGNDB.App.SmitteStop.API/Attributes/UploadKeysAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,82 @@
+                    var _ecParameters = CustomNamedCurves.GetByOid(X9ObjectIdentifiers.Prime256v1);
+                    string[] anonymousToken = authHeader.Replace("Anonymous ", string.Empty).Split(".");
+
+                    var W = _ecParameters.Curve.DecodePoint(Convert.FromBase64String(anonymousToken[0]));
```
Why not give this `W` more meaningful name e.g. `elipticCurvePoint`?

matishadow replied:
I will still argue about it.

You are using `BouncyCastle` library and you're calling `DecodePoint` method from it.  
Have you looked at the source code of it or the documentation comments of this method?
Documentation comment of this method is as follows:
```
/**
* Decode a point on this curve from its ASN.1 encoding. The different
* encodings are taken account of, including point compression for
* <code>F<sub>p</sub></code> (X9.62 s 4.2.1 pg 17).
* @return The decoded point.
*/
```
As you can see, it returns `The decoded point`, so why lose this information and rename it to `W`? 
Inside this `DecodePoint` method, local variable which is being returned is called `p` and there's no trace of `W` there.

tjesi replied:
`W` could indeed be named otherwise in the application for more readability. I would suggest `token`.

matishadow commented on DIGNDB.App.SmitteStop.API/Attributes/UploadKeysAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,82 @@
+                    string[] anonymousToken = authHeader.Replace("Anonymous ", string.Empty).Split(".");
+
+                    var W = _ecParameters.Curve.DecodePoint(Convert.FromBase64String(anonymousToken[0]));
+                    var t = Convert.FromBase64String(anonymousToken[1]);
```
Why not give this `t` more meaningful name e.g. `tokenSeed`?

jhannes replied:
Pending a decision from FHI, I'm tending to keeping the cryptographic variables as they are described in the cryptography, rather than making translations. The whole code section may be appropriate to extract into a new class (similar to JwtValidationService, but focus has to be on the app now.

matishadow replied:
I will still argue about it.

Please take a look at documentation comments of `VerifyTokenAsync` method.
```
/// <summary>
/// Used by the token verifier. It recreates the initial point from the initiator, signs it, and verifies that they are equal.
/// </summary>
/// <param name="k">The private key for the token scheme</param>
/// <param name="curve">Curve parameters</param>
/// <param name="t">Seed for the initial point chosen by the initiator</param>
/// <param name="W">Token received from the initiator</param>
/// <returns>True if the token is valid, otherwise false</returns>
```

I'm know close to nothing when it comes to cryptography, but the method you're using for token verification `VerifyTokenAsync` clearly states that `t` is `Seed for the initial point chosen by the initiator`.
Why not call it `seed` and be compliant with the used library? 

Also looking at some random RFC ([RFC5639](https://tools.ietf.org/html/rfc5639)),
in part `Appendix A.  Pseudo-Random Generation of Parameters` it says:
```
For updating the seed s of 160-bit length we use the following
   function update_seed(s):
```
and seed is called `s` there.

I'd still suggest calling it with whole words.

tjesi replied:
`tokenSeed` is a good suggestion.

matishadow commented on DIGNDB.App.SmitteStop.API/Attributes/UploadKeysAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,82 @@
+
+                    var W = _ecParameters.Curve.DecodePoint(Convert.FromBase64String(anonymousToken[0]));
+                    var t = Convert.FromBase64String(anonymousToken[1]);
+                    var k = await _privateKeyStore.GetAsync();
```
Why not give this `k` more meaningful name e.g. `privateKey`?

matishadow replied:
I don't understand why call it `k`.

Looking at `GetAsync` method:
```
public Task<BigInteger> GetAsync()
{
    var resource = $"{EmbeddedResourceConstants.ResourceBasePath}{ResourceFile}";

    var keyPair = (AsymmetricCipherKeyPair)EmbeddedPemResource.Load(resource);

    return Task.FromResult(((ECPrivateKeyParameters)keyPair.Private).D);
}
```

I'd understand calling it `D`, or even better `key`.

tjesi replied:
`privateKey` is a good suggestion.

matishadow commented:
If I was to decide I'd really consider merging this pull request.
I am aware that the Apple's/Google's solution is not perfect.
But using [a package](https://www.nuget.org/packages/AnonymousTokens.Server) which has 240 downloads for authorization purposes seems very risky to me.

jhannes commented:
> If I was to decide I'd really consider merging this pull request.
> I am aware that the Apple's/Google's solution is not perfect.
> But using [a package](https://www.nuget.org/packages/AnonymousTokens.Server) which has 240 downloads for authorization purposes seems very risky to me.

The Anonymous Tokens package was developed for this purpose only. It's could be considered part of the same pull request, but has been externalized to make the boundaries more clear. See https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Documentation/issues/4 for the full discussion

matishadow commented:
> > If I was to decide I'd really consider merging this pull request.
> > I am aware that the Apple's/Google's solution is not perfect.
> > But using [a package](https://www.nuget.org/packages/AnonymousTokens.Server) which has 240 downloads for authorization purposes seems very risky to me.
> 
> The Anonymous Tokens package was developed for this purpose only. It's could be considered part of the same pull request, but has been externalized to make the boundaries more clear. See [folkehelseinstituttet/Fhi.Smittestopp.Documentation#4](https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Documentation/issues/4) for the full discussion

Okay, if the [Fhi.Smittestopp.Verification](https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Verification) project is using this package then it's fine by me.

jhannes committed:
Merge remote-tracking branch 'origin/master' into feature/anon-tokens

jhannes committed:
Resolve issues from PR code review

jhannes commented:
I've pushed changes

matishadow commented:
Thank you for pushing changes! 

I've closed resolved comments but I'd still argue about comments with variable names. 
Could you take one more look at them please?

jhannes commented:
@matishadow Sorry about the late response. We're changing the structure for private keys in https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Verification/pull/20. Since this will affect the code under discussion, I'm waiting until I have a complete pull request

jhannes committed:
Generate private key for anonymous tokens based on certificate shared with validation server

jhannes commented:
After discussing with @martstr and @OsirisTerje I have changed the crypto names in UploadKeysAuthorizationAttribute based on a vocabulary from @martstr. The code now also supports the updated code in Fhi.Smittestopp.Validation: https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Verification/blob/d0e08858980c31f8158dd4adb74ccf3530edc274/Fhi.Smittestopp.Verification.Domain/AnonymousTokens/RollingKeyPairGenerator.cs

jhannes committed:
Add CertificateThumbPrint setting to appsettings.json

matishadow commented on DIGNDB.App.SmitteStop.API/Services/RollingKeyPairGeneratorWithLocalCertificate.cs:
```diff
@@ -0,0 +1,33 @@
+
+        private static X509Certificate2 LocateLocalCertificate(string thumbprint)
+        {
+            using (var store = new X509Store(StoreName.My, StoreLocation.LocalMachine))
```
Instead of `using` block you can use
```
using var store = new X509Store(StoreName.My, StoreLocation.LocalMachine);
```
Just a suggestion for future pull requests.

jhannes replied:
Fixed. This is neat. Makes me feel like I'm back in C++ again :-)

matishadow commented on DIGNDB.App.SmitteStop.API/Services/RollingKeyPairGeneratorWithLocalCertificate.cs:
```diff
@@ -0,0 +1,33 @@
+using System;
+using System.Security.Cryptography.X509Certificates;
+
+namespace DIGNDB.App.SmitteStop.API
```
Namespaces should match directory structure.
In general it should be `DIGNDB.App.SmitteStop.API/Services`

jhannes replied:
I fixed this now. Feeling a bit ashamed here. (In my defence, I'm used to Java where the tools fix this automatically and if you somehow make a mistake, the compiler stops you)

matishadow commented:
The crypto names are perfect now!

I've got two small comments but they're minor so I still approve with them.

jhannes committed:
Fix namespace inconsistencies

jhannes commented:
I have no idea why the build suddenly failed. There were only trivial changes and Github Actions doesn't seem to give a clarifying error message either

matishadow committed:
Merge branch 'master' into feature/anon-tokens

matishadow commented:
Merging master to your branch and rerunning the github workflow solved the build issue.
I'll investigate the issue when it comes again.

matishadow committed:
Merge branch 'master' into feature/anon-tokens

sindremb commented on DIGNDB.App.SmitteStop.API/Attributes/UploadKeysAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,72 @@
+    {
+        private readonly IJwtValidationService _jwtValidationService;
+        private readonly IAnonymousTokenKeySource _anonymousTokenKeySource;
+        private readonly ITokenVerifier _tokenVerifier = new TokenVerifier(new InMemorySeedStore());
```
The InMemorySeedStore must be replaced by a real implementation.

Looking at the existing JwtTokenReplayAttackService, it looks like we can make an implementation of ISeedStore using the JwtToken-entity and IJwtTokenRepository, as these are just an ID and an expiry. Then this implementation just needs access to a config for the duration of each key pair, and set the expiry accordingly when creating new records.

Preferably, we should either rename the JwtToken-entity, or add a new entity for the anonymous token, but it should work just keeping it as is as well. Thought @jhannes @matishadow @OsirisTerje ?

sindremb commented on DIGNDB.App.SmitteStop.API/Services/RollingKeyPairGenerator.cs:
```diff
@@ -0,0 +1,97 @@
+            throw new NotSupportedException($"Unsupported private key for certificate {certificate.Thumbprint}");
+        }
+
+        public BigInteger GetPrivateKey(string keyId)
```
The keyId provided here is never validated. At any given point in time, only 1 or 2 keyIds should be valid (current + prev during rollover). The keyId should be validated according to the key pair duration and rollover, similarly to [how it is done in the verification solution](https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Verification/blob/879dd2edd039a7d5d36c425d4b0cebb61a7985ac/Fhi.Smittestopp.Verification.Domain/AnonymousTokens/AnonymousTokenKeyStore.cs#L91). We will then need to sync config for key pair duration between the two services

sindremb replied:
The keyId provided here is never validated. At any given point in time, only 1 or 2 keyIds should be valid (current + prev during rollover). The valid keyIds should be calculated according to the key pair duration and rollover, similarly to [how it is done in the verification solution](https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Verification/blob/879dd2edd039a7d5d36c425d4b0cebb61a7985ac/Fhi.Smittestopp.Verification.Domain/AnonymousTokens/AnonymousTokenKeyStore.cs#L91). We will then need to sync config for key pair duration between the two services

OsirisTerje commented on DIGNDB.App.SmitteStop.API/Attributes/UploadKeysAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,72 @@
+            }
+        }
+
+        public void OnActionExecuted(ActionExecutedContext context)
```
Can't this be removed?

OsirisTerje replied:
And same with the "original " one

jhannes replied:
The class is a modified copy of AuthorizationAttribute.cs and I've kept everything in that class to ease comparison. I haven't studied the original class enough to know what's strictly needed or not. I recommend looking at AuthorizationAttribute and UploadKeysAuthorizationAttribute together when considering changes.

OsirisTerje commented on DIGNDB.App.SmitteStop.API/Services/RollingKeyPairGenerator.cs:
```diff
@@ -0,0 +1,97 @@
+            {
+                var privateKeyBytes = GeneratePrivateKeyBytes(keyByteSize, masterKeyBytes, keyIntervalNumber, counter);
+                privateKey = new BigInteger(privateKeyBytes);
+                counter++;
```
Add if clause to ensure this never can run infinitely, and throw an exception if above some high limit.  Just a sanity check.

sindremb replied:
Should do the same in the verification solution. The probability of having to run the loop multiple times should be low, so I guess a cap of 100 should be enough, could go even higher just to be sure..

OsirisTerje commented on DIGNDB.App.SmitteStop.API/Services/RollingKeyPairGeneratorWithLocalCertificate.cs:
```diff
@@ -0,0 +1,29 @@
+            {
+                return enumerator.Current;
+            }
+            throw new Exception($"Certificate with thumbprint {thumbprint} not found");
```
Can you add a typed custom exception instead of a generic one?

OsirisTerje committed:
Fixing PR code comments

OsirisTerje committed:
Some more code changes for the PR

Commit:
AnonymousTokenSeedStore piggybacking on the existing JwtTokenRepository
Added validation of key-ID provided by the client in the Anonymous auth header
Refactored AnonymoustokenKeySource slightly to not load the master certificate until the feature is invoked, reducing hosting environment requirements until the feature is enabled (changed the dev config to expect the same cert as Fhi.Smittestopp.Verification dev)

sindremb commented:
Made some changes intended to resolve my previous comments. However, I havent been able to verify the flow after these changes yet due to time limitations..

netcompany-martinropcke committed:
Merge branch 'master' into feature/anon-tokens

jhannes commented:
👏
